src/Evaluator.hs

```diff
@@ -38,8 +103,343 @@ data EvalError
   | EvalErrorType TypeError
   | EvalErrorCase SrcLoc Value
   | EvalErrorUser Syntax
-  deriving (Show)
+  | EvalErrorIdent Value
+  deriving (Show, Typeable)
 makePrisms ''EvalError
+instance Exception EvalError
+
+-- | The Kontinuation type. The naming convention InFoo means that the subject
+-- of evaluation in the CEK machine is Foo. For example, when the continuation
+-- is 'InArg' the subject of evaluation is the argument of the function and the
+-- continuation holds the un-evaluated function symbol in its first field.
+data Kont where
+
+  Halt :: Kont
+  -- ^ Marks the evaluator finishing
+
+  -- functions
+  InArg :: !(CoreF TypePattern ConstructorPattern Core) -> !VEnv -> !Kont -> Kont
+  -- ^ The argument is being evaluated, so hold onto the function symbol.
+  InFun :: !Value -> !VEnv -> !Kont -> Kont
+  -- ^ The function is being evaluated, so hold onto the evaluated argument.
+
+  InLetDef :: !Ident -> !Var -> !(CoreF TypePattern ConstructorPattern Core) -> !VEnv -> !Kont -> Kont
+  -- ^ Evaluating the let def
+
+  -- constructors
+  InCtor :: !Constructor -> ![CoreF TypePattern ConstructorPattern Core] -> ![Value] -> !VEnv -> !Kont -> Kont
+
+  -- Cases
+  InCaseScrut     :: ![(SyntaxPattern, Core)] -> !SrcLoc -> !VEnv -> !Kont -> Kont
+  InDataCaseScrut :: ![(ConstructorPattern, Core)] -> !SrcLoc -> !VEnv -> !Kont -> Kont
+  InTypeCaseScrut :: ![(TypePattern, Core)] -> !SrcLoc -> !VEnv -> !Kont -> Kont
+
+  -- lists
+  InConsHd :: !Core -> !(CoreF TypePattern ConstructorPattern Core) -> !VEnv -> !Kont -> Kont
+  InConsTl :: !Core -> !Syntax -> !VEnv -> !Kont -> Kont
+  InList   :: !Core -> ![Core] -> ![Syntax] -> !VEnv -> !Kont -> Kont
```

Comment:
It's a shame that you had to waste your time reimplementing those outdated direct syntax-manipulating primitives! I only ever use `open-syntax` and `close-syntax` now. We should get rid of this old stuff. In a later MR, of course.